HerPublicWebsite/Controllers/QuestionnaireController.cs

```diff
@@ -135,19 +138,76 @@ public IActionResult Address_Post(AddressViewModel viewModel)
             extraRouteValues: new Dictionary<string, object>
             {
                 { "postcode", viewModel.Postcode },
-                { "number", viewModel.BuildingNameOrNumber }
+                { "buildingNameOrNumber", viewModel.BuildingNameOrNumber }
             }
         );
 
         return RedirectToAction(forwardArgs.Action, forwardArgs.Controller, forwardArgs.Values);
     }
 
-    [HttpGet("address/{postcode}/{number}")]
-    public IActionResult SelectAddress_Get(AddressViewModel viewModel, string postcode, string buildingNameOrNumber)
+    [HttpGet("address/{postcode}/{buildingNameOrNumber}")]
+    public IActionResult SelectAddress_Get(string postcode, string buildingNameOrNumber)
+    {
+
+        //TODO (BEISHER-248): Replace with call to OS Places API to get actual results
+        var viewModel = new SelectAddressViewModel()
+        {
+            Addresses = new List<OsPlacesResult>
+            {
+                new() {
+                    Address = "82 Test Place",
+                    Uprn = "1337"
+                },
+                new ()
+                {
+                    Address = "23 Jellyifsh Place",
+                    Uprn = "420"
+                }
+            }
+        };
+
+        TempData["Addresses"] = JsonSerializer.Serialize(viewModel.Addresses);
+
+        return View("SelectAddress", viewModel);
+    }
+
+    [HttpPost("address/{postcode}/{buildingNameOrNumber}")]
+    public IActionResult SelectAddress_Post(SelectAddressViewModel viewModel, string postcode, string buildingNameOrNumber)
+    {
+        if (!ModelState.IsValid)
+        {
+            return SelectAddress_Get(postcode, buildingNameOrNumber);
+        }
+        
+        try
+        {
+            var addressResults = JsonSerializer.Deserialize<List<OsPlacesResult>>(TempData["Addresses"] as string ?? throw new InvalidOperationException());
+            var questionnaire = questionnaireService.UpdateAddress(addressResults[Convert.ToInt32(viewModel.Index)]);
+
+            var nextStep = questionFlowService.NextStep(QuestionFlowStep.SelectAddress, questionnaire);
+            return RedirectToNextStep(nextStep);
+        }
+        catch (Exception e)
+        {
+            var questionnaire = questionnaireService.GetQuestionnaire();

```

Comment:
Please add a comment to say that this should never happen, but if it does we take the user back to the page to enter their postcode and house number.
Also please add a logger error message here so that we can investigate if it ever does happen.

Reply:
Also I think conceptually we're saying "start the address lookup process again" not "go back to the previous page", so we should just go directly there with something like `return RedirectToAction(nameof(QuestionnaireController.Address_Get), "Questionnaire")